### PR TITLE
Name resolving seems overcomplicated #249

### PR DESCRIPTION
**Pull request by lburgazzoli**
- dns-srv resolver is now an optional packages
- the direct resolver is now loaded by default
- service loader is now used by default but the logic to discover resolvers is pluggable

### CODE REVIEW

**Comment by lburgazzoli** [comment]
@fanminshi do you ind reviewing it ?

**Comment by fanminshi** [comment]
@lburgazzoli sorry for the delay. I am reviewing it now.

File: jetcd-core/src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java [hunk 1]
```diff
@@ -159,7 +160,9 @@ synchronized void close() {
         .nameResolverFactory(
```
Comment (fanminshi) [inline jetcd-core/src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java]: it seems to be setting `nameResolverFactory` again is not needed. 
`defaultChannelBuilder()` has already set the `nameResolverFactory`.

File: jetcd-discovery-dns-srv/pom.xml [hunk 2]
```diff
@@ -0,0 +1,139 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (fanminshi) [inline jetcd-discovery-dns-srv/pom.xml]: I am not sure if main jetcd repo should have dns resolver implementation. the users can write one themself.
I think its better to keep code base simple and small.

File: jetcd-core/src/main/java/com/coreos/jetcd/resolver/URIResolverLoader.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  /**
+   * Creates a default implementation based on {@link ServiceLoader}.
+   *
+   * @return the default implementation.
+   */
+  static URIResolverLoader defaultLoader() {
+    return () -> {
+      List<URIResolver> resolvers = new ArrayList<>();
+
+      ServiceLoader.load(URIResolver.class).forEach(resolvers::add);
```
Comment (fanminshi) [inline jetcd-core/src/main/java/com/coreos/jetcd/resolver/URIResolverLoader.java]: How does `ServiceLoader` find  `DirectUriResolver `? I notcied that auto service for `DirectUriResolver` is removed.

**Comment by fanminshi** [comment]
@lburgazzoli sorry for the delay. I have reviewed the pr.

File: jetcd-core/src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java (hunk 1 above)
Comment (lburgazzoli) [reply jetcd-core/src/main/java/com/coreos/jetcd/internal/impl/ClientConnectionManager.java]: There is a small difference between the two: 
- `defaultChannelBuilder` uses the endpoints from the builder
- `withNewChannel` provides its own endpoint so we need to override the `nameResolverFactory`

File: jetcd-core/src/main/java/com/coreos/jetcd/resolver/URIResolverLoader.java (hunk 3 above)
Comment (lburgazzoli) [reply jetcd-core/src/main/java/com/coreos/jetcd/resolver/URIResolverLoader.java]: The DirectUriResolver is included by default, see https://github.com/lburgazzoli/coreos-jetcd/blob/github-249/jetcd-core/src/main/java/com/coreos/jetcd/resolver/SmartNameResolver.java#L63

File: jetcd-discovery-dns-srv/pom.xml (hunk 2 above)
Comment (lburgazzoli) [reply jetcd-discovery-dns-srv/pom.xml]: Well, it was already included before so I do not see the point to remove it and I think we should provide reasonable default implementation for common things like this one.

File: jetcd-core/src/main/java/com/coreos/jetcd/resolver/DirectUriResolver.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  @Override
+  public boolean supports(URI uri) {
+    if (!SCHEMES.contains(uri.getScheme())) {
+      return false;
+    }
+
+    if (!Strings.isNullOrEmpty(uri.getPath())) {
+      return false;
+    }
+
```
[the remark this answers is not in the document]
Comment (lburgazzoli) [reply jetcd-core/src/main/java/com/coreos/jetcd/resolver/DirectUriResolver.java]: fixed

File: jetcd-core/src/main/java/com/coreos/jetcd/resolver/DirectUriResolver.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          "Unsupported URI " + uri
+      );
+    }
+
+    return this.cache.computeIfAbsent(
+        uri,
+        u -> Collections.singletonList(
+            new EquivalentAddressGroup(new InetSocketAddress(
+                uri.getHost(),
+                uri.getPort() != -1 ? uri.getPort() : 2739)
```
[the remark this answers is not in the document]
Comment (lburgazzoli) [reply jetcd-core/src/main/java/com/coreos/jetcd/resolver/DirectUriResolver.java]: fixed

File: jetcd-discovery-dns-srv/pom.xml (hunk 2 above)
Comment (fanminshi) [reply jetcd-discovery-dns-srv/pom.xml]: I see.

**Comment by fanminshi** [comment]
lgtm. thanks @lburgazzoli

**Pull request merged**